File: bogo2bogo/ChocolafStyle/examples/C++/MyProjects/OpenCV/ImageViewer/ImageViewer.cpp
#include "ImageViewer.h"
#include "ImageSpinner.h"
#include "common_funcs.h"
#include "ui_ImageViewer.h"
#include <QtCore>
#include <QtGui>
#include <QtWidgets>

class VLine : public QFrame
{
public:
  VLine(QWidget *parent = nullptr) : QFrame(parent)
  {
    setFrameShape(QFrame::VLine);
    setFrameShadow(QFrame::Sunken);
  }
};

ImageViewer::ImageViewer(QWidget *parent)
    : QMainWindow(parent), image(nullptr), imageSpinner(nullptr)
{
  setWindowTitle("ImageViewer");
  scaleFactor = 1.0;
  imageLoaded = false;
  imageLabel = new QLabel("");
  imageInfoLabel = new QLabel("");
  imageCountLabel = new QLabel("");
  scaleFactorLabel = new QLabel("");
  scrollArea = new QScrollArea();

  // imageLabel->setBackgroundRole(QPalette::Base);
  imageLabel->setSizePolicy(QSizePolicy::Ignored, QSizePolicy::Ignored);
  imageLabel->setScaledContents(true);

  // scrollArea->setBackgroundRole(QPalette::Dark);
  scrollArea->setWidget(imageLabel);
  scrollArea->setVisible(false);
  setCentralWidget(scrollArea);

  createActions();
  createMenus();
  createToolbar();
  statusBar()->showMessage(
      QString("Qt %1 ImageViewer with Chocolaf theme").arg(QT_VERSION_STR));
  setupStatusBar();

  // set initial size to 3/5 of screen
  resize(QGuiApplication::primaryScreen()->availableSize() * 4 / 5);
  setWindowIcon(QIcon(":/app_icon.png")); // load an image
}

ImageViewer::~ImageViewer()
{
}

// helper function
QString getIconPath(QString baseName, bool darkTheme = false)
{
  QString iconPath =
      QString(":/%1_%2.png").arg(baseName).arg(darkTheme ? "dark" : "light");
  qDebug() << "Loading icon " << iconPath;
  return iconPath;
}

void ImageViewer::createActions()
{
  openAction = new QAction("&Open...", this);
  openAction->setShortcut(QKeySequence::Open);
  // openAction->setIcon(QIcon(":/open.png"));
  openAction->setIcon(QIcon(":/open.png"));
  openAction->setStatusTip("Open a new image file to view");
  QObject::connect(openAction, SIGNAL(triggered()), this, SLOT(open()));

  printAction = new QAction("&Print...", this);
  printAction->setShortcut(QKeySequence::Print);
  // printAction->setIcon(QIcon(":/print.png"));
  printAction->setIcon(QIcon(":/print.png"));
  printAction->setStatusTip("Print the current image");
  QObject::connect(printAction, SIGNAL(triggered()), this, SLOT(print()));
  printAction->setEnabled(false);

  exitAction = new QAction("E&xit", this);
  exitAction->setShortcut(QKeySequence("Ctrl+Q"));
  exitAction->setStatusTip("Quit the application");
  QObject::connect(exitAction, SIGNAL(triggered()), QApplication::instance(),
                   SLOT(quit()));

  zoomInAction = new QAction("Zoom &in (25%)", this);
  zoomInAction->setShortcut(QKeySequence("Ctrl++"));
  // zoomInAction->setIcon(QIcon(":/zoom_in.png"));
  zoomInAction->setIcon(QIcon(":/zoom_in.png"));
  zoomInAction->setStatusTip("Zoom into the image by 25%");
  QObject::connect(zoomInAction, SIGNAL(triggered()), this, SLOT(zoomIn()));
  zoomInAction->setEnabled(false);

  zoomOutAction = new QAction("Zoom &out (25%)", this);
  zoomOutAction->setShortcut(QKeySequence("Ctrl+-"));
  // zoomOutAction->setIcon(QIcon(":/zoom_out.png"));
  zoomOutAction->setIcon(QIcon(":/zoom_out.png"));
  zoomOutAction->setStatusTip("Zoom out of the image by 25%");
  QObject::connect(zoomOutAction, SIGNAL(triggered()), this, SLOT(zoomOut()));
  zoomOutAction->setEnabled(false);

  rotateLeftAction = new QAction("Rotate &left", this);
  rotateLeftAction->setShortcut(QKeySequence("Ctrl+<"));
  rotateLeftAction->setIcon(QIcon(":/rotate_left.png"));
  rotateLeftAction->setStatusTip("Rotate image counter-clockwise by 90 degrees");
  QObject::connect(rotateLeftAction, SIGNAL(triggered()), this, SLOT(rotateLeft()));
  rotateLeftAction->setEnabled(false);

  rotateRightAction = new QAction("Rotate &right", this);
  rotateRightAction->setShortcut(QKeySequence("Ctrl+>"));
  rotateRightAction->setIcon(QIcon(":/rotate_right.png"));
  rotateRightAction->setStatusTip("Rotate image clockwise by 90 degrees");
  QObject::connect(rotateRightAction, SIGNAL(triggered()), this, SLOT(rotateRight()));
  rotateRightAction->setEnabled(false);

  zoomNormalAction = new QAction("&Normal size", this);
  zoomNormalAction->setShortcut(QKeySequence("Ctrl+0"));
  zoomNormalAction->setStatusTip("Zoom to normal size (reset zoom)");
  QObject::connect(zoomNormalAction, SIGNAL(triggered()), this, SLOT(normalSize()));
  zoomNormalAction->setEnabled(false);

  fitToWindowAction = new QAction("Fit to &window", this);
  fitToWindowAction->setShortcut(QKeySequence("Ctrl+1"));
  // fitToWindowAction->setIcon(QIcon(":/fit_to_size.png"));
  fitToWindowAction->setIcon(QIcon(":/zoom_fit.png"));
  fitToWindowAction->setStatusTip("Fit the image to window");
  QObject::connect(fitToWindowAction, SIGNAL(triggered()), this, SLOT(fitToWindow()));
  fitToWindowAction->setEnabled(false);
  fitToWindowAction->setCheckable(true);

  prevImageAction = new QAction("&Previous image", this);
  prevImageAction->setShortcut(QKeySequence::MoveToPreviousChar); // left arrow (usually)
  // prevImageAction->setIcon(QIcon(":/previous.png"));
  prevImageAction->setIcon(QIcon(":/go_prev.png"));
  prevImageAction->setStatusTip("View previous image in folder");
  QObject::connect(prevImageAction, SIGNAL(triggered()), this, SLOT(prevImage()));
  prevImageAction->setEnabled(false);

  nextImageAction = new QAction("&Next image", this);
  nextImageAction->setShortcut(QKeySequence::MoveToNextChar); // right arrow (usually)
  // nextImageAction->setIcon(QIcon(":/next.png"));
  nextImageAction->setIcon(QIcon(":/go_next.png"));
  nextImageAction->setStatusTip("View next image in folder");
  QObject::connect(nextImageAction, SIGNAL(triggered()), this, SLOT(nextImage()));
  nextImageAction->setEnabled(false);

  aboutAction = new QAction("&About...", this);
  aboutAction->setStatusTip("Display the about information");
  QObject::connect(aboutAction, SIGNAL(triggered()), this, SLOT(about()));

  aboutQtAction = new QAction("About &Qt...", this);
  aboutQtAction->setStatusTip("Display information about the Qt library");
  QObject::connect(aboutQtAction, SIGNAL(triggered()), QApplication::instance(),
                   SLOT(aboutQt()));
}

void ImageViewer::createMenus()
{
  QMenu *fileMenu = menuBar()->addMenu("&File");
  fileMenu->addAction(openAction);
  fileMenu->addAction(printAction);
  fileMenu->addSeparator();
  fileMenu->addAction(exitAction);

  QMenu *viewMenu = menuBar()->addMenu("&View");
  viewMenu->addAction(zoomInAction);
  viewMenu->addAction(zoomOutAction);
  viewMenu->addAction(zoomNormalAction);
  viewMenu->addAction(fitToWindowAction);
  viewMenu->addSeparator();
  viewMenu->addAction(rotateLeftAction);
  viewMenu->addAction(rotateRightAction);
  viewMenu->addSeparator();
  viewMenu->addAction(prevImageAction);
  viewMenu->addAction(nextImageAction);

  QMenu *helpMenu = menuBar()->addMenu("Help");
  helpMenu->addAction(aboutAction);
  helpMenu->addAction(aboutQtAction);
}

void ImageViewer::createToolbar()
{
  QToolBar *toolBar = addToolBar("&Main");

  toolBar->addAction(openAction);
  toolBar->addAction(printAction);
  toolBar->addSeparator();
  toolBar->addAction(zoomInAction);
  toolBar->addAction(zoomOutAction);
  toolBar->addAction(fitToWindowAction);
  toolBar->addAction(rotateLeftAction);
  toolBar->addAction(rotateRightAction);
  toolBar->addSeparator();
  toolBar->addAction(prevImageAction);
  toolBar->addAction(nextImageAction);
}

void ImageViewer::updateActions()
{
  fitToWindowAction->setEnabled(imageLoaded);
  zoomInAction->setEnabled(!fitToWindowAction->isChecked());
  zoomOutAction->setEnabled(!fitToWindowAction->isChecked());
  zoomNormalAction->setEnabled(!fitToWindowAction->isChecked());
  prevImageAction->setEnabled(imageLoaded);
  nextImageAction->setEnabled(imageLoaded);
}

void ImageViewer::setupStatusBar()
{
  // statusBar()->reformat();
  statusBar()->setStyleSheet("QStatusBar::item {border: none;}");
  // statusBar()->addPermanentWidget(new VLine());
  statusBar()->addPermanentWidget(imageInfoLabel);
  statusBar()->addPermanentWidget(imageCountLabel);
  statusBar()->addPermanentWidget(scaleFactorLabel);
}

void ImageViewer::updateStatusBar()
{
  if (imageSpinner) {
#ifdef USING_QT6
    QImage image = imageLabel->pixmap().toImage();
#else
    QImage image = imageLabel->pixmap()->toImage();
#endif
    auto imageInfoText = QString("%1 x %2 %3")
                             .arg(image.width())
                             .arg(image.height())
                             .arg(image.isGrayscale() ? "grayscale" : "color");
    imageInfoLabel->setText(imageInfoText);
    auto sizeWidth = QString("%1").arg(imageSpinner->size()).length();
    QString status = QString("%1 of %2 images")
                         .arg(imageSpinner->currIndex() + 1, sizeWidth)
                         .arg(imageSpinner->size(), sizeWidth);
    qDebug() << status;
    imageCountLabel->setText(status);
    if (fitToWindowAction->isChecked()) {
      scaleFactorLabel->setText(QString("Zoom: %1").arg("Fit"));
    } else {
      auto currZoomFactor = int(scaleFactor * 100);
      scaleFactorLabel->setText(QString("Zoom: %1%").arg(currZoomFactor));
    }
  }
}

void ImageViewer::scaleImage(double factor /*= -1*/)
{
  // NOTE: factor == -1 is used to scale a newly loaded image to same
  // scaleFactor as last loaded image
  if (factor != -1)
    scaleFactor *= factor;
  imageLabel->resize(scaleFactor * imageLabel->pixmap(Qt::ReturnByValue).size());

  adjustScrollBar(scrollArea->horizontalScrollBar(), factor);
  adjustScrollBar(scrollArea->verticalScrollBar(), factor);

  zoomInAction->setEnabled(scaleFactor < 5.0);
  zoomOutAction->setEnabled(scaleFactor > 0.10);
}

void ImageViewer::adjustScrollBar(QScrollBar *scrollBar, double factor)
{
  scrollBar->setValue(
      int(factor * scrollBar->value() + ((factor - 1) * scrollBar->pageStep() / 2)));
}

bool ImageViewer::loadImage(const QString &imagePath)
{
  QImageReader reader(imagePath);
  reader.setAutoTransform(true);
  const QImage newImage = reader.read();
  if (newImage.isNull()) {
    imageLoaded = false;
    QMessageBox::information(this, "ImageViewer",
                             QString("FATAL: could not load image %1").arg(imagePath));
    return false;
  }

  // scaleFactor = 1.0;
  imageLoaded = true;
  scrollArea->setVisible(true);
  imageLabel->setPixmap(QPixmap::fromImage(newImage));
  imageLabel->adjustSize();
  scaleImage();

  QString title;
  QTextStream ostr(&title);
  ostr << "Qt " << QT_VERSION_STR << " ImageViewer: " << imagePath;
  setWindowTitle(title);

  // add ImageSpinner
  delete imageSpinner;
  imageSpinner = new ImageSpinner(imagePath);
  updateStatusBar();

  return true;
}

void ImageViewer::initializeFileDialog(QFileDialog &dialog,
                                       QFileDialog::AcceptMode acceptMode)
{
  const QStringList picsLocation =
      QStandardPaths::standardLocations(QStandardPaths::PicturesLocation);
  dialog.setDirectory(picsLocation.isEmpty() ? QDir::currentPath() : picsLocation.last());

  QStringList mimeTypeFilters;
  const QByteArrayList supportedMimeTypes = (acceptMode == QFileDialog::AcceptOpen)
                                                ? QImageReader::supportedMimeTypes()
                                                : QImageWriter::supportedMimeTypes();
  for (const QByteArray &mimeTypeName : supportedMimeTypes)
    mimeTypeFilters.append(mimeTypeName);
  mimeTypeFilters.sort();
  dialog.setMimeTypeFilters(mimeTypeFilters);
  dialog.selectMimeTypeFilter("image/jpeg");
  dialog.setAcceptMode(acceptMode);
  if (acceptMode == QFileDialog::AcceptSave)
    dialog.setDefaultSuffix("jpg");
}

void ImageViewer::open()
{
  QFileDialog dialog(this, "Open Image");
  initializeFileDialog(dialog, QFileDialog::AcceptOpen);

  if (dialog.exec() == QFileDialog::Accepted &&
      loadImage(dialog.selectedFiles().constFirst()))
    updateActions();
}

void ImageViewer::print()
{
  QMessageBox::information(this, "ImageViewer", "TODO: Implementation for Print Image");
}

void ImageViewer::zoomIn()
{
  scaleImage(1.25); // zoom in by 25%
  updateStatusBar();
}

void ImageViewer::zoomOut()
{
  scaleImage(0.75); // zoom out by 25%
  updateStatusBar();
}

void ImageViewer::normalSize()
{
  imageLabel->adjustSize();
  scaleFactor = 1.0;
  updateStatusBar();
}

void ImageViewer::fitToWindow()
{
  bool fitToWindow = fitToWindowAction->isChecked();
  scrollArea->setWidgetResizable(fitToWindow);
  if (!fitToWindow)
    normalSize();
  updateActions();
  updateStatusBar();
}

void ImageViewer::rotateLeft()
{
  // TODO
}

void ImageViewer::rotateRight()
{
  // TODO
}

void ImageViewer::prevImage()
{
  if (!imageSpinner->atFirst()) {
    QString imagePath = imageSpinner->prevImage();
    if (loadImage(imagePath))
      updateActions();
  } else {
    if (imageSpinner->atFirst())
      QMessageBox::information(this, "ImageViewer", "Displaying first image in folder!");
  }
}

void ImageViewer::nextImage()
{
  if (!imageSpinner->atLast()) {
    QString imagePath = imageSpinner->nextImage();
    if (loadImage(imagePath))
      updateActions();
  } else {
    if (imageSpinner->atLast())
      QMessageBox::information(this, "ImageViewer", "Displaying last image in folder!");
  }
}

void ImageViewer::about()
{
  QString str = QString("<b>Image Viewer</b> application to view images on desktop.<br/>"
                        "Created with Qt %1 and Chocolaf theme<br/><br/>"
                        "Developed by Manish Bhobe<br/>"
                        "Free to use, but use at your own risk!!")
                    .arg(QT_VERSION_STR);
  QMessageBox::about(this, tr("About Image Viewer"), str);
}

File: bogo2bogo/ChocolafStyle/examples/C++/MyProjects/OpenCV/ImageViewer/ImageViewer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# ImageViewer.py: Image viewer application with PyQt
import os
import pathlib
import sys
import cv2
from argparse import ArgumentParser

from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

from chocolaf.utils.chocolafapp import ChocolafApp
from ImageSpinner import ImageSpinner
import ImageViewer_rc


class ImageViewer(QMainWindow):
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)
        self.imageLoaded = False
        self.imageLabel = QLabel("")
        self.imageCountLabel = QLabel("")
        self.scaleFactorLabel = QLabel("")
        self.imageInfoLabel = QLabel("")
        self.scrollArea = QScrollArea()
        self.cv2image = None        # OpenCV Mat
        self.image = None           # QImage
        self.scaleFactor = 1.0
        self.firstDialog = True
        self.imageSpinner = None
        self.setupUi()

    def setupUi(self):
        self.imageLabel.setBackgroundRole(QPalette.Base)
        self.imageLabel.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Ignored)
        self.imageLabel.setScaledContents(True)

        self.scrollArea.setBackgroundRole(QPalette.Dark)
        self.scrollArea.setWidget(self.imageLabel)
        self.scrollArea.setVisible(False)
        self.setCentralWidget(self.scrollArea)

        self.createActions()
        self.createMenu()
        self.createToolbar()
        # and status bar
        self.statusBar().showMessage(
            f"Image Viewer: developed with PyQt {PYQT_VERSION_STR} by Manish Bhobe")
        self.setupStatusBar()

        self.resize(QGuiApplication.primaryScreen().availableSize() * (4 / 5))
        self.setWindowIcon(QIcon(":/app_icon.png"))

    def createActions(self):
        # open image
        self.openAction = QAction("&Open...", self)
        self.openAction.setShortcut(QKeySequence.New)
        self.openAction.setIcon(QIcon(":/open.png"))
        self.openAction.setStatusTip("Open a new image file to view")
        self.openAction.triggered.connect(self.open)

        # print image
        self.printAction = QAction("&Print...", self)
        self.printAction.setShortcut(QKeySequence.Print)
        self.printAction.setIcon(QIcon(":/print.png"))
        self.printAction.setStatusTip("Print the current image")
        self.printAction.triggered.connect(self.print)
        self.printAction.setEnabled(False)

        # exit
        self.exitAction = QAction("E&xit", self)
        self.exitAction.setShortcut(QKeySequence("Ctrl+Q"))
        self.exitAction.setStatusTip("Exit the application")
        self.exitAction.triggered.connect(QApplication.instance().quit)

        # View category...
        self.zoomInAction = QAction("Zoom &in (25%)", self)
        self.zoomInAction.setShortcut(QKeySequence("Ctrl++"))
        self.zoomInAction.setIcon(QIcon(":/zoom_in.png"))
        self.zoomInAction.setStatusTip("Zoom into the image by 25%")
        self.zoomInAction.triggered.connect(self.zoomIn)
        self.zoomInAction.setEnabled(False)

        self.zoomOutAction = QAction("Zoom &out (25%)", self)
        self.zoomOutAction.setShortcut(QKeySequence("Ctrl++"))
        self.zoomOutAction.setIcon(QIcon(":/zoom_out.png"))
        self.zoomOutAction.setStatusTip("Zoom out of the image by 25%")
        self.zoomOutAction.triggered.connect(self.zoomOut)
        self.zoomOutAction.setEnabled(False)

        self.zoomNormalAction = QAction("&Normal size", self)
        self.zoomNormalAction.setShortcut(QKeySequence("Ctrl+0"))
        self.zoomNormalAction.setStatusTip("Zoom to normal size")
        self.zoomNormalAction.triggered.connect(self.zoomNormal)
        self.zoomNormalAction.setEnabled(False)

        self.fitToWindowAction = QAction("Fit to &window", self)
        self.fitToWindowAction.setShortcut(QKeySequence("Ctrl+1"))
        self.fitToWindowAction.setIcon(QIcon(":/zoom_fit.png"))
        self.fitToWindowAction.setStatusTip("Fit image to size of window")
        self.fitToWindowAction.triggered.connect(self.fitToWindow)
        self.fitToWindowAction.setEnabled(False)
        self.fitToWindowAction.setCheckable(True)

        self.prevImageAction = QAction("&Previous Image", self)
        self.prevImageAction.setShortcut(QKeySequence.MoveToPreviousChar)
        self.prevImageAction.setIcon(QIcon(":/go_prev.png"))
        self.prevImageAction.setStatusTip("View previous image in folder")
        self.prevImageAction.triggered.connect(self.prevImage)
        self.prevImageAction.setEnabled(False)

        self.nextImageAction = QAction("&Next Image", self)
        self.nextImageAction.setShortcut(QKeySequence.MoveToNextChar)
        self.nextImageAction.setIcon(QIcon(":/go_next.png"))
        self.nextImageAction.setStatusTip("View next image in folder")
        self.nextImageAction.triggered.connect(self.nextImage)
        self.nextImageAction.setEnabled(False)

        # Help category
        self.aboutAction = QAction("&About...", self)
        self.aboutAction.setStatusTip("Display about application information")
        self.aboutAction.triggered.connect(self.about)

        self.aboutQtAction = QAction("About &Qt...", self)
        self.aboutQtAction.setStatusTip(
            "Display information about Qt library being used")
        self.aboutQtAction.triggered.connect(QApplication.instance().aboutQt)

    def createMenu(self):
        # file menu
        fileMenu = self.menuBar().addMenu("&File")
        fileMenu.addAction(self.openAction)
        fileMenu.addAction(self.printAction)
        fileMenu.addSeparator()
        fileMenu.addAction(self.exitAction)

        # view menu
        viewMenu = self.menuBar().addMenu("View")
        viewMenu.addAction(self.zoomInAction)
        viewMenu.addAction(self.zoomOutAction)
        viewMenu.addAction(self.zoomNormalAction)
        viewMenu.addSeparator()
        viewMenu.addAction(self.fitToWindowAction)
        viewMenu.addSeparator()
        viewMenu.addAction(self.prevImageAction)
        viewMenu.addAction(self.nextImageAction)

        # help menu
        helpMenu = self.menuBar().addMenu("Help")
        helpMenu.addAction(self.aboutAction)
        helpMenu.addAction(self.aboutQtAction)

    def createToolbar(self):
        toolBar = self.addToolBar("Main")
        toolBar.addAction(self.openAction)
        toolBar.addAction(self.printAction)
        toolBar.addSeparator()
        toolBar.addAction(self.zoomInAction)
        toolBar.addAction(self.zoomOutAction)
        toolBar.addAction(self.fitToWindowAction)
        toolBar.addAction(self.prevImageAction)
        toolBar.addAction(self.nextImageAction)

    def updateActions(self):
        self.fitToWindowAction.setEnabled(not (self.image is None))
        self.zoomInAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.zoomOutAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.zoomNormalAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.prevImageAction.setEnabled(not (self.image is None))
        self.nextImageAction.setEnabled(not (self.image is None))

    def setupStatusBar(self):
        self.statusBar().setStyleSheet("QStatusBar::item {border: none;}")
        self.statusBar().addPermanentWidget(self.imageInfoLabel)
        self.statusBar().addPermanentWidget(self.imageCountLabel)
        self.statusBar().addPermanentWidget(self.scaleFactorLabel)

    def updateStatusBar(self):
        """ update the various permanent widgets on the status bar """
        if self.imageSpinner is not None:
            imageInfoText = f"{self.image.width():4d} x {self.image.height():4d} {'grayscale' if self.image.isGrayscale() else 'color'}"
            self.imageInfoLabel.setText(imageInfoText)
            selImageText = f"{self.imageSpinner.currIndex+1:3d} of {self.imageSpinner.size():3d} images"
            self.imageCountLabel.setText(selImageText)
            scaleFactorText = "Zoom: Fit" if self.fitToWindowAction.isChecked() \
                else f"Zoom: {int(self.scaleFactor * 100)} %"
            self.scaleFactorLabel.setText(scaleFactorText)

    def openCV2QImage(self, cv2image):
        """ convert an OpenCV2 image read using cv2.imread to QImage """
        height, width, channels = cv2image.shape
        bytes_per_line = width * channels
        qimage = QImage(cv2image, width, height, bytes_per_line, QImage.Format_RGB888)
        return qimage

    def displayImageInfo(self):
        if self.image.isNull():
            return
        print(f"Image info -> width: {self.image.width()} - height: {self.image.height()} " +
              f"- bits/pixel: {self.image.depth()} - color: {not self.image.isGrayscale()}")

    def scaleImage(self, factor=-1):
        """scale image to a certain scaling factor. Default value of -1 is 
           only used to scale a newly loaded image to same scale factor as prev image
        """
        if factor != -1:
            self.scaleFactor *= factor
        self.imageLabel.resize(self.scaleFactor * self.imageLabel.pixmap().size())
        self.adjustScrollbar(self.scrollArea.horizontalScrollBar(), factor)
        self.adjustScrollbar(self.scrollArea.verticalScrollBar(), factor)
        self.zoomInAction.setEnabled(self.scaleFactor < 5.0)
        self.zoomOutAction.setEnabled(self.scaleFactor > 0.10)
        #print(f"Scalefactor = {self.scaleFactor}")

    def adjustScrollbar(self, scrollBar: QScrollBar, factor):
        scrollBar.setValue(int(factor * scrollBar.value() +
                               ((factor - 1) * scrollBar.pageStep() / 2)))

    def initOpenDialog(self, dialog: QFileDialog, acceptMode: QFileDialog.AcceptMode):
        picLocations = QStandardPaths.standardLocations(QStandardPaths.PicturesLocation)
        # print(f"Standard pic locations: {picLocations}")
        dialog.setDirectory(QDir.currentPath if len(picLocations) == 0 else picLocations[-1])
        supportedMimeTypes = (QImageReader.supportedMimeTypes() if acceptMode == QFileDialog.AcceptMode.AcceptOpen
                              else QImageWriter.supportedMimeTypes())
        mimeTypeFilters = [str(mimeTypeName, 'utf-8') for mimeTypeName in supportedMimeTypes]
        mimeTypeFilters = sorted(mimeTypeFilters)
        dialog.setMimeTypeFilters(mimeTypeFilters)
        dialog.selectMimeTypeFilter("image/jpeg")
        dialog.setAcceptMode(acceptMode)
        if (acceptMode == QFileDialog.AcceptMode.AcceptSave):
            dialog.setDefaultSuffix("jpg")

    def loadImage(self, imagePath):
        assert os.path.exists(imagePath), f"FATAL: could not load file {filePath}"
        # reader = QImageReader(imagePath)
        # reader.setAutoTransform(True)
        # self.image = reader.read()
        # if self.image is None:
        #     QMessageBox.error(self, "ImageViewer", f"Could not load {imagePath}")
        #     return False

        self.cv2image = cv2.imread(imagePath)
        if self.cv2image.size == 0:
            QMessageBox.error(self, "ImageEditor", f"Could not load {imagePath}")
            return False
        else:
            self.cv2image = cv2.cvtColor(self.cv2image, cv2.COLOR_BGR2RGB)

        self.image = self.openCV2QImage(self.cv2image)

        # self.scaleFactor = 1.0
        self.scrollArea.setVisible(True)

        # self.imageLabel.setPixmap(QPixmap.fromImage(self.image))
        # self.imageLabel.adjustSize()
        # # self.fitToWindow()
        # self.scaleImage()
        self.showImage(self.image)

        self.setWindowTitle(f"PyQt {PYQT_VERSION_STR} Image Viewer: {imagePath}")

        # image spinner
        if (self.imageSpinner is not None):
            del self.imageSpinner
        self.imageSpinner = ImageSpinner(imagePath)
        self.updateStatusBar()

        return True

    def showImage(self, image: QImage):
        self.imageLabel.setPixmap(QPixmap.fromImage(image))
        self.imageLabel.adjustSize()
        # self.fitToWindow()
        self.scaleImage()

    def open(self):
        openDialog = QFileDialog(self, "Open Image")
        self.initOpenDialog(openDialog, QFileDialog.AcceptMode.AcceptOpen)
        if openDialog.exec():
            filePath = list(openDialog.selectedFiles())[-1]
            if self.loadImage(filePath):
                self.updateActions()
                self.displayImageInfo()

    def print(self):
        QMessageBox.information(self, "ImageViewer",
                                "This is the 'print' action handler - yet to be implemented")

    def zoomIn(self):
        self.scaleImage(1.25)
        self.updateStatusBar()

    def zoomOut(self):
        self.scaleImage(0.75)
        self.updateStatusBar()

    def zoomNormal(self):
        self.imageLabel.adjustSize()
        self.scaleFactor = 1.0
        self.updateStatusBar()

    def fitToWindow(self):
        fitToWindow = self.fitToWindowAction.isChecked()
        self.scrollArea.setWidgetResizable(fitToWindow)
        if not fitToWindow:
            self.zoomNormal()
        self.updateActions()

    def prevImage(self):
        if not self.imageSpinner.atFirstPath():
            assert (self.imageSpinner is not None)
            imagePath = self.imageSpinner.prevImagePath()
            if (self.loadImage(imagePath)):
                self.updateActions()
        else:
            QMessageBox.information(self, "ImageViewer", "Displaying first image in folder!")

    def nextImage(self):
        if not self.imageSpinner.atLastPath():
            assert (self.imageSpinner is not None)
            imagePath = self.imageSpinner.nextImagePath()
            if (self.loadImage(imagePath)):
                self.updateActions()
        else:
            QMessageBox.information(self, "ImageViewer", "Displaying last image in folder!")

    def about(self):
        QMessageBox.about(self, "About Image Viewer",
                          f"<b>Image Viewer</b> application to view images on desktop.<br/>"
                          f"Developed with PyQt {PYQT_VERSION_STR} and Chocolaf theme<br/><br/>"
                          f"Version 1.0, by Manish Bhobe<br/>"
                          f"Free to use, but use at your own risk!!")


def main():
    # app = QApplication(sys.argv)
    # # app.setFont(QApplication.font("QMenu"))
    # app.setStyle("Fusion")
    # # palSwitcher = utils.PaletteSwitcher(app)
    #
    # if darkdetect.isDark():
    #     ThemeSetter.setDarkTheme(app)
    # else:
    #     ThemeSetter.setLightTheme(app)
    ap = ArgumentParser()
    ap.add_argument("-i", "--image", required=False,
                    help="Full path to image")
    args = vars(ap.parse_args())
    app = ChocolafApp(sys.argv)
    app.setStyle("Chocolaf")

    w = ImageViewer()
    w.setWindowTitle(f"PyQt {PYQT_VERSION_STR} Image Viewer")
    if args['image'] is not None:
        # check if image path provided
        if os.path.exists(args['image']):
            w.loadImage(args['image'])
        else:
            print(f"WARNING: {args['image']} - path does not exist!")
    w.show()

    sys.exit(app.exec())


if __name__ == '__main__':
    main()
